core/src/main/scala/kafka/server/KafkaConfig.scala

```diff
@@ -1949,6 +1951,17 @@ class KafkaConfig(val props: java.util.Map[_, _], doLog: Boolean, dynamicConfigO
       )
     }
 
+    val voterIds: Set[Integer] = if (usesSelfManagedQuorum) RaftConfig.parseVoterConnections(quorumVoters).asScala.keySet.toSet else Set.empty
+    if (voterIds.nonEmpty) {
+      if (processRoles.contains(ControllerRole)) {
+        // Ensure that controllers use their node.id as a voter in controller.quorum.voters 
+        require(voterIds.contains(nodeId), s"If ${KafkaConfig.ProcessRolesProp} contains the 'controller' role, the node id $nodeId must be included in the set of voters ${RaftConfig.QUORUM_VOTERS_CONFIG}=$voterIds")
+      } else {
+        // Ensure that the broker's node.id is not an id in controller.quorum.voters
+        require(!voterIds.contains(nodeId), s"If ${KafkaConfig.ProcessRolesProp} does not contain the 'controller' role, the node id $nodeId must not be included in the set of voters ${RaftConfig.QUORUM_VOTERS_CONFIG}=$voterIds")
+      }
+    }
+
```

Comment:
I think we need an additional check that the voter list is not empty if `process.roles` is set. Otherwise we will skip the validation that `node.id` is included in the voter list when we are a controller.

Reply:
This code already checks for that on line 1955.

Reply:
Right, but if the voters list _is_ empty, and we are in self-managed mode, I believe we should throw an error.

Reply:
Good catch